source/client/stream_decoder.cc

```diff
@@ -19,27 +20,37 @@ void StreamDecoder::decodeHeaders(Envoy::Http::HeaderMapPtr&& headers, bool end_
 
 void StreamDecoder::decodeData(Envoy::Buffer::Instance&, bool end_stream) {
   ASSERT(!complete_);
-  complete_ = end_stream;
-  if (complete_) {
+  if (!have_first_body_byte_) {
+    have_first_body_byte_ = true;
+    milestone_tracker_.markMilestone(milestone_first_body_byte_);
+  }
+  if (end_stream) {
+    complete_ = end_stream;
+    milestone_tracker_.markMilestone(milestone_last_body_byte_);
     onComplete(true);
   }
 }
 
 void StreamDecoder::decodeTrailers(Envoy::Http::HeaderMapPtr&&) {
   ASSERT(!complete_);
+  milestone_tracker_.markMilestone(milestone_trailer_);
   complete_ = true;
   onComplete(true);
 }
 
 void StreamDecoder::onComplete(bool success) {
-  if (success && measure_latencies_) {
-    latency_statistic_.addValue((time_source_.monotonicTime() - request_start_).count());
-  }
   ASSERT(!success || complete_);
+  milestone_tracker_.markMilestone(milestone_response_complete_);
```

Comment:
Maybe add some comments here and in `onPoolReady` below about what stats are being tracked?